### ISSUE TITLE
Racy deployment creation

### ISSUE THREAD

**Issue by bobrik** [mesosphere/marathon#1407]
I noticed that some deployments from [ansible role](https://github.com/Topface/ansible-marathon_app) report as finished immediately (under ~200ms) while marathon ui thinks that they are running.

I wrote a simple script that creates an app with `PUT /v2/apps`, gets deployment id and checks if deployment appears immediately in `GET /v2/deployments`. I was't able to replicate results over ssh tunnel, but in datacenter it revealed racy behavior.

Here's my test script:

```sh
#!/bin/sh

set -e

MARATHON=$1
if [ -z "${MARATHON}" ]; then
    echo "usage: <marathon uri> <app id> <app definition>"
    exit 1
fi

MARATHON_APP=$2
if [ -z "${MARATHON_APP}" ]; then
    echo "usage: <marathon uri> <app definition>"
    exit 1
fi

MARATHON_APP_ID=$(echo "${MARATHON_APP}" | jq -r .id)

echo "creating ${MARATHON_APP_ID}"

MARATHON_DEPLOYMENT_ID=$(curl -s -H "Content-type: application/json" -X PUT "${MARATHON}/v2/apps/${MARATHON_APP_ID}" -d "${MARATHON_APP}" | jq -r .deploymentId)

echo "got deployment id: ${MARATHON_DEPLOYMENT_ID}"

curl -s "${MARATHON}/v2/deployments" | jq .

sleep 1

curl -s "${MARATHON}/v2/deployments" | jq .
```

Here's the output:

```
# /tmp/test.sh http://web488:8080 '{"id":"/racy/test1","cmd":"sleep 60","cpus":0.1,"mem":16}'
creating /racy/test1
got deployment id: 917eea45-fc73-4951-89a4-0d5b17c761f4
[]
[
  {
    "currentStep": 1,
    "currentActions": [
      {
        "action": "RestartApplication",
        "app": "/racy/test1"
      }
    ],
    "affectedApps": [
      "/racy/test1"
    ],
    "version": "2015-04-17T13:24:50.601Z",
    "id": "917eea45-fc73-4951-89a4-0d5b17c761f4",
    "totalSteps": 1,
    "steps": [
      [
        {
          "action": "RestartApplication",
          "app": "/racy/test1"
        }
      ]
    ]
  }
]
```

I'm on 0.8.1-SNAPSHOT, not very current one.

**Comment by drexin** [comment]
How lon does it stay running in the UI? The UI only reloads every 5 seconds, so a certain amount of delay is expected. I don't see a real problem with this, but maybe I'm missing something.

**Comment by bobrik** [comment]
Real problem is: I can't reliably report that deployment is done, because API immediately thinks that there is not deployment at all (see first reply in script output).

Here's what I always expect to see:

1. I PUT an application.
2. Deployment appears in API immediately or doesn't appear at all.
3. After deployment is finished, it never appears again.

Here's what I see:

1. I PUT an application.
2. Deployment is missing in API.
3. Deployment appears.

UI and its refresh rate is not my concern.

**Comment by drexin** [comment]
Ok, sorry for the confusion. We'll have a look.

**Comment by bobrik** [comment]
Thanks, looking forward to it

**Comment by kolloch** [comment]
We issue the App writes to Zookeeper before returning the deployment now (we did not do that before 0.11).

But it turns out that Zookeeper writes are not immediately visible for subsequent reads. "Consistency Guarantees - ZooKeeper is a high performance, scalable service. Both reads and write operations are designed to be fast, though reads are faster than writes. The reason for this is that in the case of reads, ZooKeeper can serve older data, which in turn is due to ZooKeeper's consistency guarantees." (https://zookeeper.apache.org/doc/trunk/zookeeperProgrammers.html)

This should probably be handled:

* by serving read requests from memory (this is consistent because we only have one leader)
* using the sync request when a new leader starts up to make sure that it sees all prior writes

We will first implement this for tasks: https://github.com/mesosphere/marathon/issues/462

**Comment by kolloch** [comment]
Hey @bobrik, I decided to promote this issue to a more general issue and close this.

 #2299

**Comment by meichstedt** [comment]
Close – will be fixed via #2299
